File: static/js/calculator2.js
document.addEventListener('DOMContentLoaded', function () {
  const form = document.getElementById('calculatorForm');
  const resultsDiv = document.getElementById('results');

  form.addEventListener('submit', function (e) {
    e.preventDefault();

    const composition = {
      carbonCombustible:
        parseFloat(document.getElementById('carbonCombustible').value) || 0,
      hydrogenCombustible:
        parseFloat(document.getElementById('hydrogenCombustible').value) || 0,
      oxygenCombustible:
        parseFloat(document.getElementById('oxygenCombustible').value) || 0,
      sulfurCombustible:
        parseFloat(document.getElementById('sulfurCombustible').value) || 0,
      vanadiumCombustible:
        parseFloat(document.getElementById('vanadiumCombustible').value) || 0,
      moistureContent:
        parseFloat(document.getElementById('moistureContent').value) || 0,
      ashDry: parseFloat(document.getElementById('ashDry').value) || 0,
      heatingValueCombustible:
        parseFloat(document.getElementById('heatingValueCombustible').value) ||
        0,
    };

    fetch('/calculator2', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(composition),
    })
      .then(response => {
        if (!response.ok) {
          throw new Error('Помилка сервера');
        }
        return response.json();
      })
      .then(data => {
        displayResults(data);
      })
      .catch(error => {
        alert('Помилка: ' + error.message);
      });
  });

  function displayResults(results) {
    // склад робочої маси
    const workingCompositionDiv = document.getElementById('workingComposition');
    workingCompositionDiv.innerHTML = '';

    Object.entries(results.workingComposition).forEach(([component, value]) => {
      const row = document.createElement('div');
      row.className = 'row mb-2';

      const labelCol = document.createElement('div');
      labelCol.className = 'col-8';
      labelCol.textContent = component;

      const valueCol = document.createElement('div');
      valueCol.className = 'col-4 text-end';

      // для ванадію використовуємо мг/кг, для інших - відсотки
      if (component.includes('V')) {
        valueCol.textContent = `${value.toFixed(2)} мг/кг`;
      } else {
        valueCol.textContent = `${value.toFixed(2)}%`;
      }

      row.appendChild(labelCol);
      row.appendChild(valueCol);
      workingCompositionDiv.appendChild(row);
    });

    // нижча теплота згоряння
    const heatingValueDiv = document.getElementById('workingHeatingValue');
    heatingValueDiv.textContent = `${results.workingHeatingValue.toFixed(
      2,
    )} МДж/кг`;

    resultsDiv.style.display = 'block';
  }
});

File: calculator2.go
package main

import (
	"encoding/json"
	"math"
	"net/http"
)

// компоненти мазуту
type MazutComposition struct {
	CarbonCombustible     float64 `json:"carbonCombustible"`
	HydrogenCombustible   float64 `json:"hydrogenCombustible"`
	OxygenCombustible     float64 `json:"oxygenCombustible"`
	SulfurCombustible     float64 `json:"sulfurCombustible"`
	VanadiumCombustible   float64 `json:"vanadiumCombustible"`
	MoistureContent       float64 `json:"moistureContent"`
	AshDry                float64 `json:"ashDry"`
	HeatingValueCombustible float64 `json:"heatingValueCombustible"`
}

// результати розрахунків
type MazutResults struct {
	WorkingComposition  map[string]float64 `json:"workingComposition"`
	WorkingHeatingValue float64            `json:"workingHeatingValue"`
}

func calculateMazutResults(composition MazutComposition) MazutResults {
	// формула перерахунку складу палива
	conversionFactor := (100.0 - composition.MoistureContent - composition.AshDry) / 100.0

	// перерахунок для кожного компонента
	workingComposition := map[string]float64{
		"C^P": composition.CarbonCombustible * conversionFactor,
		"H^P": composition.HydrogenCombustible * conversionFactor,
		"O^P": composition.OxygenCombustible * conversionFactor,
		"S^P": composition.SulfurCombustible * conversionFactor,
		// віднімання - отримання маси без вологи (суха маса)
		// оскільки ванадій вказується відносно маси без вологи
		// ділення - отримання відсотку
		"V^P": composition.VanadiumCombustible * (100.0 - composition.MoistureContent) / 100.0,
		// віднімання - отримання відсотка маси без вологи
		// оскільки зола вказується відносно маси без вологи
		// ділення - отримання відсотку
		"A^P": composition.AshDry * (100.0 - composition.MoistureContent) / 100.0,
	}

	// розрахунок нижчої теплоти згорання
	workingHeatingValue := composition.HeatingValueCombustible*
		(100.0-composition.MoistureContent-composition.AshDry)/100.0 -
		0.025*composition.MoistureContent

	// Округлюємо результати до 2 знаків після коми
	for k, v := range workingComposition {
		workingComposition[k] = math.Round(v*100) / 100
	}

	return MazutResults{
		WorkingComposition:  workingComposition,
		WorkingHeatingValue: math.Round(workingHeatingValue*100) / 100,
	}
}

// обробка запитів до калькулятора мазуту
func HandleCalculator2(w http.ResponseWriter, r *http.Request) {
	if r.Method == "POST" {
		var composition MazutComposition
		err := json.NewDecoder(r.Body).Decode(&composition)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		results := calculateMazutResults(composition)

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(results)
		return
	}

	http.ServeFile(w, r, "templates/calculator2.html")
}

// налаштування маршрутів для калькулятора 2
func SetupCalculator2Routes() {
	http.HandleFunc("/calculator2", HandleCalculator2)
}